go/adbc/adbc.go

```diff
@@ -0,0 +1,378 @@
+	//
+	// 1. The column's ordinal position in the table (starting from 1).
+	// 2. Database-specific description of the column.
+	// 3. Optional Value. Should be null if not supported by the driver.
+	//	  xdbc_values are meant to provide JDBC/ODBC-compatible metadata
+	//		in an agnostic manner.
+	//
+	// CONSTRAINT_SCHEMA is a Struct with the fields:
+	//
+	//		Field Name									| Field Type					| Comments
+	//		----------------------------|---------------------|---------
+	//		constraint_name							| utf8								|
+	//		constraint_type							| utf8 not null				| (1)
+	//		constraint_column_names			| list<utf8> not null | (2)
+	//		constraint_column_usage			| list<USAGE_SCHEMA>	| (3)
+	//
+	// 1. One of 'CHECK', 'FOREIGN KEY', 'PRIMARY KEY', or 'UNIQUE'.
+	// 2. The columns on the current table that are constrained, in order.
+	// 3. For FOREIGN KEY only, the referenced table and columns.
+	//
+	// USAGE_SCHEMA is a Struct with fields:
+	//
+	//		Field Name									|	Field Type
+	//		----------------------------|----------------------------
+	//		fk_catalog									| utf8
+	//		fk_db_schema								| utf8
+	//		fk_table										| utf8 not null
+	//		fk_column_name							| utf8 not null
+	//
+	// For the parameters: If nil is passed, then that parameter will not
+	// be filtered by at all. If an empty string, then only objects without
+	// that property (ie: catalog or db schema) will be returned.
+	//
+	// tableName and columnName must be either nil (do not filter by
+	// table name or column name) or non-empty.
+	//
+	// All non-empty, non-nil strings should be a search pattern (as described
+	// earlier).
+	GetObjects(ctx context.Context, depth ObjectDepth, catalog, dbSchema, tableName, columnName *string, tableType []string) (array.RecordReader, error)
+
+	GetTableSchema(ctx context.Context, catalog, dbSchema *string, tableName string) (*arrow.Schema, error)
+
+	// GetTableTypes returns a list of the table types in the database.
+	//
+	// The result is an arrow dataset with the following schema:
+	//
+	//		Field Name			| Field Type
+	//		----------------|--------------
+	//		table_type			| utf8 not null
+	//
+	GetTableTypes(context.Context) (array.RecordReader, error)
+	Commit(context.Context) error
+	Rollback(context.Context) error
+	NewStatement(options map[string]string) (Statement, error)
+	Close() error
+	ReadPartition(ctx context.Context, serializedPartition []byte) (array.RecordReader, error)
+}
+
+// PostInitOptions is an optional interface which can be implemented by
+// drivers which allow modifying and setting options after initializing
+// a connection or statement.
+type PostInitOptions interface {
+	SetOption(key, value string) error
+}
+
+type Partitions struct {
+	NumPartitions uint64
+	PartitionIDs  [][]byte
+}
+
+type Statement interface {
+	Close() error
+	SetSqlQuery(query string) error
+	ExecuteQuery(context.Context) (array.RecordReader, int64, error)
+	ExecuteUpdate(context.Context) (int64, error)
```

Comment:
I guess do you want to consolidate these as well?

Reply:
I suppose since the reader isn't an out parameter, though, you don't get the differentiation - so maybe not

Reply:
Yea, i was keeping the differentiation by not having the RecordReader output